scheduler/scheduler_system.go

```diff
@@ -276,86 +279,177 @@ func (s *SystemScheduler) computeJobAllocs() error {
-		s.planner.ServersMeetMinimumVersion(minVersionMaxClientDisconnect, true))
+	reconciliationResult := nr.Compute(s.job, s.nodes, s.notReadyNodes, tainted, s.feasibleNodesForTG,
+		live, term, s.planner.ServersMeetMinimumVersion(minVersionMaxClientDisconnect, true))
 	if s.logger.IsDebug() {
-		s.logger.Debug("reconciled current state with desired state", r.Fields()...)
+		s.logger.Debug("reconciled current state with desired state", reconciliationResult.Fields()...)
 	}
 
-	// Add the deployment changes to the plan
-	s.plan.Deployment = nr.DeploymentCurrent
-	s.plan.DeploymentUpdates = nr.DeploymentUpdates
-
 	// Update the stored deployment
 	if nr.DeploymentCurrent != nil {
 		s.deployment = nr.DeploymentCurrent
 	}
 
 	// Add all the allocs to stop
-	for _, e := range r.Stop {
+	for _, e := range reconciliationResult.Stop {
 		s.plan.AppendStoppedAlloc(e.Alloc, sstructs.StatusAllocNotNeeded, "", "")
 	}
 
 	// Add all the allocs to migrate
-	for _, e := range r.Migrate {
+	for _, e := range reconciliationResult.Migrate {
 		s.plan.AppendStoppedAlloc(e.Alloc, sstructs.StatusAllocNodeTainted, "", "")
 	}
 
 	// Lost allocations should be transitioned to desired status stop and client
 	// status lost.
-	for _, e := range r.Lost {
+	for _, e := range reconciliationResult.Lost {
 		s.plan.AppendStoppedAlloc(e.Alloc, sstructs.StatusAllocLost, structs.AllocClientStatusLost, "")
 	}
 
-	for _, e := range r.Disconnecting {
+	for _, e := range reconciliationResult.Disconnecting {
 		s.plan.AppendUnknownAlloc(e.Alloc)
 	}
 
 	allocExistsForTaskGroup := map[string]bool{}
 	// Attempt to do the upgrades in place.
 	// Reconnecting allocations need to be updated to persists alloc state
 	// changes.
-	updates := make([]reconciler.AllocTuple, 0, len(r.Update)+len(r.Reconnecting))
-	updates = append(updates, r.Update...)
-	updates = append(updates, r.Reconnecting...)
+	updates := make([]reconciler.AllocTuple, 0, len(reconciliationResult.Update)+len(reconciliationResult.Reconnecting))
+	updates = append(updates, reconciliationResult.Update...)
+	updates = append(updates, reconciliationResult.Reconnecting...)
 	destructiveUpdates, inplaceUpdates := inplaceUpdate(s.ctx, s.eval, s.job, s.stack, updates)
-	r.Update = destructiveUpdates
+	reconciliationResult.Update = destructiveUpdates
 
 	for _, inplaceUpdate := range inplaceUpdates {
 		allocExistsForTaskGroup[inplaceUpdate.TaskGroup.Name] = true
 	}
 
 	s.planAnnotations = &structs.PlanAnnotations{
-		DesiredTGUpdates: desiredUpdates(r, inplaceUpdates, destructiveUpdates),
+		DesiredTGUpdates: desiredUpdates(reconciliationResult, inplaceUpdates, destructiveUpdates),
 	}
 
 	// Treat non in-place updates as an eviction and new placement, which will
 	// be limited by max_parallel
-	s.limitReached = evictAndPlace(s.ctx, s.job, r, sstructs.StatusAllocUpdating)
+	s.limitReached = evictAndPlace(s.ctx, s.job, reconciliationResult, sstructs.StatusAllocUpdating)
```

Comment:
We probably want to keep `evictAndPlace` before `computePlacements` for sake of efficiency, but currently it includes limiting the canaries, which we then repeat after `computePlacements`. I think we want to bypass it entirely if we're canarying, so that we can do the equivalent logic after `computePlacements`.